chamge/amqp-arbiter-backend.c

```diff
@@ -32,65 +33,294 @@ G_DEFINE_TYPE (ChamgeAmqpArbiterBackend, chamge_amqp_arbiter_backend, CHAMGE_TYP
+      connection_info.host, connection_info.port,
+      connection_info.vhost, connection_info.user, connection_info.password);
+  return CHAMGE_RETURN_OK;
+}
 
-  struct amqp_connection_info amqp_c_info;
-  amqp_bytes_t queue;
+static ChamgeReturn
+_amqp_declare_queue (const amqp_connection_state_t conn,
+    const gchar * queue_name, amqp_bytes_t * queue)
+{
   amqp_queue_declare_ok_t *amqp_r = NULL;
+  gchar *response_queue_name = NULL;
+
+  if (queue_name == NULL) {
+    amqp_r = amqp_queue_declare (conn, 1, amqp_empty_bytes, 0, 0, 0, 1,
+        amqp_empty_table);
+  } else {
+    amqp_r =
+        amqp_queue_declare (conn, 1, amqp_cstring_bytes (queue_name), 0, 0, 0,
+        1, amqp_empty_table);
+  }
+  if (amqp_r == NULL) {
+    g_error ("Can not declare a new queue or queue does not exist...");
+    return CHAMGE_RETURN_FAIL;
+  }
+  amqp_get_rpc_reply (conn);
 
-  ChamgeAmqpArbiterBackend *self =
-      CHAMGE_AMQP_ARBITER_BACKEND (arbiter_backend);
+  g_assert_nonnull (amqp_r->queue.bytes);
+  response_queue_name = g_strdup (amqp_r->queue.bytes);
+  *queue = amqp_cstring_bytes (response_queue_name);
 
-  amqp_uri = g_settings_get_string (self->settings, "amqp-uri");
-  g_assert_nonnull (amqp_uri);
+  g_debug ("create private reply_to queue : name -> %s",
+      (gchar *) queue->bytes);
+  return CHAMGE_RETURN_OK;
+}
 
-  amqp_channel = g_settings_get_int (self->settings, "amqp-channel");
 
-  amqp_parse_url (amqp_uri, &amqp_c_info);
+static ChamgeReturn
+_amqp_rpc_subscribe (amqp_connection_state_t amqp_conn, guint channel,
+    const gchar * exchange_name, const gchar * queue_name)
+{
+  amqp_bytes_t amqp_queue = amqp_cstring_bytes (queue_name);
+  amqp_bytes_t amqp_listen_queue;
+
+  g_return_val_if_fail (amqp_conn, CHAMGE_RETURN_FAIL);
+  if (_amqp_declare_queue (amqp_conn, queue_name, &amqp_listen_queue) !=
+      CHAMGE_RETURN_OK) {
+    return CHAMGE_RETURN_FAIL;
+  }
+  if (g_strcmp0 (queue_name, amqp_listen_queue.bytes) != 0) {
+    g_error ("declare required queue name (%s) != declared queue name (%s)",
+        queue_name, (char *) amqp_listen_queue.bytes);
+    return CHAMGE_RETURN_FAIL;
+  }
+  amqp_bytes_free (amqp_listen_queue);
+
+  amqp_queue_bind (amqp_conn, channel, amqp_queue,
+      amqp_cstring_bytes (exchange_name), amqp_queue, amqp_empty_table);
+
+  g_debug ("binding a queue(%s) (exchange: %s, bind_key: %s)",
+      queue_name, exchange_name, queue_name);
+
+  amqp_basic_consume (amqp_conn, channel, amqp_queue,
+      amqp_empty_bytes, 0, 0, 0, amqp_empty_table);
+  g_return_val_if_fail ((amqp_get_rpc_reply (amqp_conn)).reply_type ==
+      AMQP_RESPONSE_NORMAL, CHAMGE_RETURN_FAIL);
+
+  return CHAMGE_RETURN_OK;
+}
+
+static ChamgeReturn
+_amqp_property_set (amqp_basic_properties_t * const props,
+    const guint correlation_id, const gchar * reply_queue_name)
+{
+  gchar *correlation_id_str = NULL;
+  gchar *content_type_str = g_strdup (DEFAULT_CONTENT_TYPE);
```

Comment:
`content_type_str` can be leaked if the pointer isn't properly handled before return.
The below coding pattern is a way to avoid leak.

```
   g_autofree gchar *content_type_str = g_strdup(..);
   ...
   props->content_type = amqp_cstring_bytes  (g_steal_pointer (&content_type_str));
```